### ISSUE TITLE
Unclear how to access outputs/intermediate outputs correctly

### ISSUE THREAD

**Issue by andrewhong5297** [Unity-Technologies/barracuda-release#54]
I have tried following the guide to access intermeditate outputs with "additionalOutputs" but I keep getting just the layer name and inputs outputted to me instead of the actual output. 

I want access the string "After Softmax" that is outputted when verbose is true, but haven't found a way to do so. The final model output is "Identity" which I'm not sure how to access either. (should be output of array size 2, predicting dog or cat. 

Thanks in advance!

![debug](https://user-images.githubusercontent.com/47720952/83893475-fbe1f080-a71d-11ea-8233-f5fe058661b0.PNG)

**Comment by AlexRibard** [comment]
Hi @andrewhong5297,
When you specify `additionalOutputs` when creating the worker, you will need to call
```
outputName = "After SoftMax";
var additionalOutputs = new string[] {outputName }
m_Worker = WorkerFactory.CreateWorker(<WorkerFactory.Type>, m_RuntimeModel, additionalOutputs);
...
var output = worker.PeekOutput(outputName);
```
output will be a Tensor storing the data of that intermediate node.

Be sure the name of the desired output matches the ones in the onnx file (you can either check in the inspector window of your model, or with model.layers)

Hope that helps!

**Comment by andrewhong5297** [comment]
Thanks for getting back to me. "After Softmax" isn't a layer though, it's seems to be an output from verbose that I don't know how to call. All I know is it comes after the Layer referenced first in the image above. 

When I use "After Softmax" as part of the additionalOutputs, I get a "KeyNotFoundException". 
![image](https://user-images.githubusercontent.com/47720952/83896915-df948280-a722-11ea-81b9-9b1626ba80d4.png)

The onnx model looks like this, its just a slight variation of the mobilenetv2 model.
![image](https://user-images.githubusercontent.com/47720952/83896961-ef13cb80-a722-11ea-89c1-d06f89873ee6.png)

**Comment by AlexRibard** [comment]
Can you share the onnx model directly?

Yeah it's probably not the node name, but some debug text. You need to match the name in the inspector/onnx file.
Judging from your screenshot, it would seem that the softmax layer is called
`StatefulPartitionedCall/sequential/dense/Softmax`
Try with that.

**Comment by andrewhong5297** [comment]
When I try that and print it in the debug log with peekoutput I get: 

(`StatefulPartitionedCall/sequential/dense/Softmax_out` (1, 1, 1, 2), alloc: Unity.Barracuda.DefaultTensorAllocator, onDevice:(GPU:#1606555904 (1, 14, 14, 64) buffer: UnityEngine.ComputeBuffer created at: ))
UnityEngine.Debug:LogWarning(Object)
TFtest:ProcessImage() (at Assets/Scripts/ML Stuff/TFtest.cs:92)
TFtest:Update() (at Assets/Scripts/ML Stuff/TFtest.cs:36)

But I can't get the softmax values. It doesn't seem like I can share the onnx model on this comment since the file type isn't supported.

**Comment by AlexRibard** [comment]
I think github accepts `zip` files or you can share a link to a google drive.

Ok so unfortunely the ToString() method of a `Tensor` does not print the internal data, especially if said Tensor is on the GPU. It meant more for a debug output more than an data introspection method. Sorry for the potential confusion.

You can fetch the Tensor data as follows:
```
float value = output[0];
float[] values = output.AsFloats();
```

**Comment by andrewhong5297** [comment]
https://drive.google.com/file/d/18PyctQZSHM2yUJgxqwxhF8gExIeQcAYD/view?usp=sharing

Okay let me try that out

**Comment by andrewhong5297** [comment]
It is working now! Thank you so much Alex :)

**Comment by AlexRibard** [comment]
Awesome! I will update the docs appropriately :)

**Comment by andrewhong5297** [comment]
Hey Alex! Sorry to bother again, have you ever tried the tiny yolo v2 or similar yolo models before? My outputs are all NaN, model is attached here:

https://drive.google.com/file/d/1w5aVtLVo1_O98BD8oxi2O7PBuC6ZTBHl/view?usp=sharing

![image](https://user-images.githubusercontent.com/47720952/83958909-6477bd00-a845-11ea-89b7-e0c58483cdda.png)

**Comment by AlexRibard** [comment]
I exectuted your model with a random input with `onnx-runtime` and I am also getting NANs for every output.
So it would seem to come from the model itself (most likely some of the Batchnorms)
What results are you getting with TF?

Side comment, your model contains a lot of unecessary `Transpose`. When exporting from TF, use the flag `inputs_as_nchw=...` to make sure it optimizes said Transposes

**Comment by andrewhong5297** [comment]
I'm unfamiliar with that argument, what should I set that flag = to? This is just the model I directly pulled from yolo's website, maybe it's giving issues because it is so old. Would barracuda work on yolov3 onnx files?

**Comment by andrewhong5297** [comment]
I'm trying the yolov2 (not tiny) instead, and getting new error. I'm  happy to open this as a new issue if that's easier/better

![image](https://user-images.githubusercontent.com/47720952/83973899-7a25ca80-a8b7-11ea-94a2-55050126ee61.png)


https://drive.google.com/file/d/1NNqkkzyOhV4tMjh-OXQCad5TgJKr0Qok/view?usp=sharing

This is the output of the .h5 model:
![image](https://user-images.githubusercontent.com/47720952/83973841-08e61780-a8b7-11ea-9032-2c2c63134db1.png)

**Comment by andrewhong5297** [comment]
seems related to these prior issues, but those were in v0.6 so not sure if they have been addressed now. 

https://github.com/Unity-Technologies/barracuda-release/issues/9

https://github.com/Unity-Technologies/ml-agents/issues/2030

It seems mantasp has gotten it to work before though I'm not sure what he means in this comment:
![image](https://user-images.githubusercontent.com/47720952/83981019-297a9580-a8e8-11ea-95d4-39dd5ed3b742.png)

https://github.com/Unity-Technologies/ml-agents/issues/2188

**Comment by andrewhong5297** [comment]
for anyone else, I found a working example of tiny yolov2 in barracuda 

https://classifai.net/blog/tensorflow-onnx-unity/

**Comment by AlexRibard** [comment]
@andrewhong5297
* NAN issue: the problem is the following 
In the first BatchNorm, the first variance value is negative (-0.126) which results in a NAN.
How that negative number got there, that I don't know...

* yolov2:
Imports and runs correctly (validated against onnxruntime) on 1.0

* Tensorflow internal data layout is NCHW, and ONNX is NHWC. (https://carlthome.github.io/posts/nhwc-vs.-nchw%20/)
When converting with tf2onnx, the default behavior of the importer is to keep the layer data the same. So it inserts a Transpose (NHWC -> NCHW) before layers. Which is not perf friendly.
So you need to tell tf2onnx that you actually want things to be swizzled correctly. For that you tell it which inputs are actually NCHW:
```
    inputs = ['input_1:0'] # for the yolomodel you provided
    g = tf2onnx.tfonnx.process_tf_graph(sess.graph,input_names=inputs, output_names=outputs, inputs_as_nchw=inputs)
```

* yolov3:
Atm the end of the network is problematic
![image](https://user-images.githubusercontent.com/12990800/84056172-d9e9a780-a983-11ea-9163-478870fe7a86.png)
We don't support 5d tensors so the first reshape doesn't import.
So unfortunately at this point you'd need to strip out that part and do that manually.

Hope that helps!
Let me know
